src/python/pants/backend/python/macros/poetry_requirements_test.py

```diff
@@ -0,0 +1,393 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from textwrap import dedent
+from typing import Any, Dict, Iterable
+
+import pytest
+from packaging.version import Version
+from pkg_resources import Requirement
+
+from pants.backend.python.macros.poetry_requirements import (
+    PoetryRequirements,
+    get_max_caret,
+    get_max_tilde,
+    handle_dict_attr,
+    parse_pyproject_toml,
+    parse_single_dependency,
+    parse_str_version,
+)
+from pants.backend.python.target_types import PythonRequirementLibrary, PythonRequirementsFile
+from pants.base.specs import AddressSpecs, DescendantAddresses, FilesystemSpecs, Specs
+from pants.engine.addresses import Address
+from pants.engine.internals.scheduler import ExecutionError
+from pants.engine.target import Targets
+from pants.testutil.rule_runner import QueryRule, RuleRunner
+
+
+@pytest.mark.parametrize(
+    "test, exp",
+    [
+        ("1.0.0-rc0", "2.0.0"),
+        ("1.2.3.dev0", "2.0.0"),
+        ("1.2.3-dev0", "2.0.0"),
+        ("1.2.3dev0", "2.0.0"),
+        ("1.2.3", "2.0.0"),
+        ("1.2", "2.0.0"),
+        ("1", "2.0.0"),
+        ("0.2.3", "0.3.0"),
+        ("0.0.3", "0.0.4"),
+        ("0.0", "0.1.0"),
+        ("0", "1.0.0"),
+    ],
+)
+def test_caret(test, exp) -> None:
+    version = Version(test)
+    assert get_max_caret(version) == exp
+
+
+@pytest.mark.parametrize(
+    "test, exp",
+    [
+        ("1.2.3", "1.3.0"),
+        ("1.2", "1.3.0"),
+        ("1", "2.0.0"),
+        ("0", "1.0.0"),
+        ("1.2.3.rc1", "1.3.0"),
+        ("1.2.3rc1", "1.3.0"),
+        ("1.2.3-rc1", "1.3.0"),
+    ],
+)
+def test_max_tilde(test, exp) -> None:
+    version = Version(test)
+    assert get_max_tilde(version) == exp
+
+
+@pytest.mark.parametrize(
+    "test, exp",
+    [
+        ("~1.0.0rc0", ">=1.0.0,<1.1.0"),
+        ("^1.0.0rc0", ">=1.0.0,<2.0.0"),
```

Comment:
I think that these are wrong, should likely be `>=1.0.0rc0` for both. In the production code, that can be fixed by restoring your old `min_version` code as the raw string w/ prefix chopped off.

Reply:
The caret/tilde operators are somewhat variable; I have this as agreeing with NPM at the moment...I wasn't 100% sure how to check this with Poetry

Reply:
Yeah, looks like `.public` does the trick!

Reply:
Bump on this and fixing the tests. Lgtm otherwise :)

Reply:
Oh oops -- changed now; thought I was supposed to leave to match NPM.

Reply:
Thanks! I'm not sure why NPM does that, this seems like more correct behavior